Marlin/Cartridge.cpp

```diff
@@ -0,0 +1,237 @@
+//====================== Private Functions Prototypes =======================
+//===========================================================================
+
+static void cartridgeAbsentUpdate(unsigned int cartNumber);
+static void cartridgePresentUpdate(unsigned int cartNumber);
+static bool cartridgesRemovedCheck(void);
+static bool cartridgesPresentCheck(void);
+
+//===========================================================================
+//============================ Public Functions =============================
+//===========================================================================
+
+ /**
+ * Check to see if cartridges are present or absent. Flags internally if 
+ * one has been removed, or clears the removed flag if it's present. 
+ * The status of cartridge removal can be found with 
+ * CartridgeRemoved()
+ */
+void UpdateCartridgeStatus(void) {
+    // Cartridge zero is pulled low by default
+    if (READ(CART0_SIG2_PIN) == HIGH) {
+        cartridgePresentUpdate(0);
+    } else {
+        cartridgeAbsentUpdate(0);
+    }
+    // Cartridge one is pulled high by default
+    if (READ(CART1_SIG2_PIN) == LOW) {
+        cartridgePresentUpdate(1);
+    } else {
+        cartridgeAbsentUpdate(1);
+    }
+}
+
+/**
+ * This function checks to see if a cartridge has been removed from the
+ * system, allowing us to make judgement calls for error reporting. This
+ * information is updated by calling UpdateCartridgeStatus(). Will also 
+ * return true if no cartridges are present. This version includes hysteresis
+ * so that when the cartridge is re-inserted, we don't immediately throw a 
+ * temperature error (we'll read max temp when a cartridge is removed and
+ * this may not update immediately when we call this function)
+ * @returns    Returns true if a cartridge has been removed, or no 
+ *             cartridge is present WITH HYSTERESIS
+ */
+bool CartridgeRemoved(void) {
+    bool returnValue = false;
+    static unsigned int cartridgeRemovalHysteresis = 0;
+    bool removedCondition = (!cartridgePresent[0] || !cartridgesPresentCheck());
+
+    // If a cartridge is seen to be removed, set the hysteresis counter.
+    // If the appropriate cartridges are absent but we don't see that a
+    // cartridge has been removed, we started up with it missing and will also
+    // mark it as removed.
+    if (cartridgesRemovedCheck()) {
+        cartridgeRemovalHysteresis = CARTRIDGE_REMOVAL_HYSTERESIS_COUNT;
+    } else if (removedCondition) {
+        cartridgeRemovalHysteresis = CARTRIDGE_REMOVAL_HYSTERESIS_COUNT;
+    }
+
+    // We have this hysteresis here to accomodate putting the cartridge
+    // back in. Without it, Marlin will recognize the cartridge has been
+    // reinserted before the temperature updates from its maximum value and
+    // will throw a high temperature error.
+    if (cartridgeRemovalHysteresis > 0) {
+        cartridgeRemovalHysteresis--;
```

Comment:
Sure. Hysteresis might be a bit of a misnomer, but the idea here is that the function will continue returning that a cartridge has been removed for several calls after the cartridge has been placed. This is basically to prevent a temperature error, as this check seems to update quicker than our temperature reading. Before it, when a cartridge was placed in, we'd report that we'd seen a cartridge but our temperature would still read 1100 for several passes through which would result in a max temp error as soon as the cartridge was inserted.